casper/src/main/scala/coop/rchain/casper/Estimator.scala

```diff
@@ -21,37 +21,53 @@ object Estimator {
 
   def tips[F[_]: Monad: BlockStore](
       blockDag: BlockDagRepresentation[F],
-      lastFinalizedBlockHash: BlockHash
+      genesis: BlockMessage
   ): F[IndexedSeq[BlockMessage]] =
     for {
       latestMessageHashes <- blockDag.latestMessageHashes
-      result              <- Estimator.tips[F](blockDag, lastFinalizedBlockHash, latestMessageHashes)
+      result              <- Estimator.tips[F](blockDag, genesis, latestMessageHashes)
     } yield result
 
   /**
     * When the BlockDag has an empty latestMessages, tips will return IndexedSeq(genesis)
     *
-    * TODO: If the base block between the main parent and a secondary parent are more than
-    * X blocks deep from the main parent, ignore. Additionally, the last finalized block must
-    * be deeper than X blocks from the tip. This allows different validators to have
-    * different last finalized blocks and still come up with the same estimator tips for a block.
+    * TODO: Remove lastFinalizedBlockHash in follow up PR
     */
   def tips[F[_]: Monad: BlockStore](
       blockDag: BlockDagRepresentation[F],
-      lastFinalizedBlockHash: BlockHash,
+      genesis: BlockMessage,
       latestMessagesHashes: Map[Validator, BlockHash]
   ): F[IndexedSeq[BlockMessage]] =
     for {
-      scoresMap <- buildScoresMap(blockDag, latestMessagesHashes, lastFinalizedBlockHash)
+      gca       <- calculateGca(blockDag, genesis, latestMessagesHashes)
+      scoresMap <- buildScoresMap(blockDag, latestMessagesHashes, gca)
       sortedChildrenHash <- sortChildren(
-                             List(lastFinalizedBlockHash),
+                             List(gca),
                              blockDag,
                              scoresMap
                            )
       maybeSortedChildren <- sortedChildrenHash.traverse(BlockStore[F].get)
       sortedChildren      = maybeSortedChildren.flatten.toVector
     } yield sortedChildren
 
+  private def calculateGca[F[_]: Monad: BlockStore](
+      blockDag: BlockDagRepresentation[F],
+      genesis: BlockMessage,
+      latestMessagesHashes: Map[Validator, BlockHash]
+  ): F[BlockHash] =
+    for {
+      latestMessages <- latestMessagesHashes.values.toList
```

Comment:
toStream instead of toList will avoid the creation of an intermediary list

Reply:
Will do